capirca/lib/versa.py

```diff
@@ -0,0 +1,877 @@
+      predef_str = 'predefined-services-list ['
+      for predef in self.term.versa_application:
+        predef_str = predef_str + ' ' + predef
+      predef_str = predef_str + ' ];'
+      mstr.append(predef_str)
+    if len(mstr) > 0 :
+      services= Tree('services', mstr)
+      services.AddParent(p_node)
+
+  def BuildTermDscp(self, p_node):
+    """Build Term dscp"""
+    valid_dscp = []
+
+    for dscp in self.term.dscp_match:
+      if int(dscp) >= 0 and int(dscp) <= 63:
+        valid_dscp.append(dscp)
+      else:
+        pass
+    if valid_dscp:
+      valid_dscp_str = 'dscp [ ' + ' '.join(valid_dscp) + ' ];'
+      dscp = Tree('', valid_dscp_str)
+      dscp.AddParent(p_node)
+
+  def BuildTermLogging(self, p_node):
+    """Build the Term Logging """
+    set_term = Tree('set')
+    set_term.AddParent(p_node)
+    action_val = self.term.action[0]
+    if action_val == 'accept':
+      action_val = 'allow'
+    action=Tree('', 'action ' + action_val + ';')
+    action.AddParent(set_term)
+    log_event = ''
+    if not self.term.logging:
+      log_event = 'start'
+    else:
+      log_event = self.term.logging[0]
+    lef = Tree('lef', 'event '+ log_event + ';')
+    lef.AddParent(set_term)
+
+  def BuildTerm(self, p_node):
+    """Build the Term Tree"""
+
+    access_pn=Tree('access-policy ' + self.term.name )
+    access_pn.AddParent(p_node)
+    if self.verbose and self.term.comment:
+      comm=Tree('', '/* ' + self.term.comment[0] + ' */')
+      comm.AddParent(access_pn)
+
+    rule_match =Tree('match')
+    rule_match.AddParent(access_pn)
+
+    if self.from_zone:
+      self.BuildTermZone(rule_match, 'src')
+
+    if self.to_zone:
+      self.BuildTermZone(rule_match, 'dest')
+
+    if self.term.versa_application or self.app:
+      self.BuildTermApp(rule_match)
+
+    if self.term.dscp_match:
+      self.BuildTermDscp(rule_match)
+
+    if self.term.action:
+      self.BuildTermLogging(access_pn)
+
+    #print("\n".join(set_term.PrintTree()))
+
+
+
+class Versa(aclgenerator.ACLGenerator):
+  """Versa rendering class.
+
+     This class takes a policy object and renders the output into a syntax
+     which is understood by Versa firewalls.
+
+     Args:
+       pol: policy.Policy object
+  """
+
+  _PLATFORM = 'versa'
+  SUFFIX = '.vsp'
+  _SUPPORTED_AF = set(('inet', 'inet6', 'mixed'))
+  _ZONE_ADDR_BOOK = 'address-book-zone'
+  _GLOBAL_ADDR_BOOK = 'address-book-global'
+  _ADDRESSBOOK_TYPES = set((_ZONE_ADDR_BOOK, _GLOBAL_ADDR_BOOK))
+  _NOVERBOSE = 'noverbose'
+  _SUPPORTED_TARGET_OPTIONS = set((_ZONE_ADDR_BOOK,
+                                   _GLOBAL_ADDR_BOOK,
+                                   _NOVERBOSE))
+  _VERSA_SUPPORTED_TARGET_OPTIONS = set(('template',
+                                         'tenant',
+                                         'policy'))
+  _VERSA_UNSUPPORTED_TERM_OPTIONS = set(('icmp_type',
+                                         'stateless_reply',
```

Comment:
stateless_reply fields are used to skip the reply flows if the firewall is stateful. Since Versa is stateful, we can remove this from here.

Reply:
Not being used. Deleted